docs/options.md
# Options

The following settings are considered self-explainatory:

-   *Break On Single Newline*
-   *Live Update*
-   *Open Preview In Split Pane*
-   *Enable Math Rendering By Default*
-   *Use GitHub.com style*
-   *Enable Pandoc Parser*

### A note on Pandoc and *Break On Single Newline*

This option does not affect Pandoc parser.

You should be able to coerce Pandoc into breaking lines on single newline by setting the appropriate `hard_line_breaks` flag (See the [Formats](http://pandoc.org/MANUAL.html#general-options) and [Non-pandoc extensions](http://pandoc.org/MANUAL.html#non-pandoc-extensions) of the Pandoc manual)

TL;DR: `hard_line_breaks` extension should do it.

![Example of 'Pandoc Options: Markdown Flavor' option value: markdown+hard_line_breaks ](https://cloud.githubusercontent.com/assets/1201875/22458834/2c4f8764-e796-11e6-9411-140ce4b57438.png)

## Pandoc Options

If *Enable Pandoc Parser* is enabled, you need to specify:

-   *Path*: Full path to pandoc executable, `which pandoc` on unix

-   *Commandline Arguments*:
    The commandline arguments for pandoc need to be comma separated.
    Please note that arguments which take an argument like `--filter`
    should be defined as `--filter=/path/to/exe` instead of
    `--filter /path/to/exe`

-   *Markdown Flavor*:
    Here you can define your Markdown flavor, as defined [here](http://pandoc.org/README.html#pandocs-markdown)
    Please note, that raw tex does not work when rendering markdown to html
    (see [issue #46][issue-46]).

-   *Citations*:
    Whether to use pandoc [citation][pandoc-cit] functionality.

### Pandoc citations

If you have enabled Pandoc to render the markdown preview then you can enable
citation replacement by enabling **Pandoc Options: Citations** on the MPP
settings page.

MPP will now search for any file named *bibliography.bib* and *custom.csl*
from the markdown's directory up. The first files that are matching will be
used for Pandocs citations. You can change the filenames it is searching for
by changing the options **Bibliography (bibfile)** and **Bibliography Style
(cslfile)** on the settings page.

You can set a fallback bibfile and cslfile if the current repository
has none of it's own.

Here is a small example how it works:

```` text
./
├── bibliography.bib     <-- will be used by README.md
├── custom.csl           <-- will be used by README.md & RANDOM.md
├── src
│   ├── bibliography.bib <-- will be used by RANDOM.md
│   ├── otherbib.bib     <-- will not be used as filename does not match
│   └── md
│       └── RANDOM.md
└── README.md
````

Effectively the arguments `--csl=./custom.csl --bibliography=./bibliography.bib`
are used for `./README.md` and `--csl=./custom.csl
--bibliography=./src/bibliography.bib` for `./src/md/RANDOM.md`

[issue-46]: https://github.com/Galadirith/markdown-preview-plus/issues/46#issuecomment-124324926
[pandoc-cit]: http://pandoc.org/README.html#citations


CHANGELOG.md
## 2.5.1
* Deactivate `markdown-preview` package on our activation
* Disambiguate serializer name from `markdown-preview` package
* Use custom view tag to differentiate ourselves from `markdown-preview`

## 2.5.0
* Bundle our fork of markdown-it-math
* Configurable math delimiters (#327)
* Ignore fence name case (#338)
* Make file extensions that are considered Markdown configurable (#335)
* Get rid of fs-plus
* Emoji support (#244, #332)
* Fix deserialization
* Make styles a bit more specific
* Bind editor-related commands on text editor instead of workspace
* CheckBox lists (#186)
* Rewrite in TypeScript (#323)

## 2.4.16
* Tentative fix for #316

## 2.4.15
* Remove dependency on cheerio
* Revert "Properly handle link\/image urlencoding" (problems on Win32)
* Revert "Undefined\/null-aware en-\/decodeURI" (problems on Win32)

## 2.4.14
* Add dedicated 'Pandoc Options: Filters' option

## 2.4.13
* Undefined\/null-aware en-\/decodeURI
* Properly handle link\/image urlencoding

## 2.4.12
* Do not break image data URIs

## 2.4.11
* [Pandoc] Remove dependency on pdc; saner Pandoc error handling

## 2.4.10
* Fix image update bug
* Atom 1.19 compatibility
* Use travis trusty env
* Specs fix for newer Atom
* Update README.md (#297) (Alan Yee)
* Typo fix (#281) (Kevin Murray)

## 2.4.9
* Atom community move (#276)
* Fix image-update fail on win32

## 2.4.8
* Do not re-render preview on image change

## 2.4.7
* Fix #273

## 2.4.6
* Fix resource leak in image-watcher
* [CHANGELOG] Markdown formatting fixes

## 2.4.5
* Update CHANGELOG
* Update CHANGELOG
* Merge branch '2.4-branch'
* Merge branch 'lierdakil/pandoc-run-path' into 2.4-branch
* Set pandoc run path to file dir

## 2.4.4
* Prepare 2.4.4 release
* Update CHANGELOG
* Merge branch 'lierdakil/pandoc-run-path' into 2.4-branch
* Merge branch '2.4-branch'

## 2.4.3
* Prepare 2.4.3 release
* Change MathJax cdn to https
* Merge pull request #264 from Galadirith/lierdakil/pandoc-native-code-style
* Merge branch 'rkichenama-master'
* update menu for split direction, closes Galadirith/markdown-preview-plus#248 (Richard Kichenama)
* Set pandoc run path to file dir

## 2.4.2
* Prepare 2.4.2 release
* Update CHANGELOG
* Merge pull request #265 from Galadirith/lierdakil/fix-citeproc-error-display
* Merge branch 'html-highlights'
* Fix 'pandoc: [object Object]: does not exist'
* Support pandoc's native code style in html export
* pandoc native code style option
* Fix code highlighting with saveAs
* Fix pandoc source code higlighting
* SaveAs: MathJax should use request's schema
* No mathjax success notifications if not in devMode
* Merge pull request #258 from Galadirith/lierdakil/update-specs
* [Spec] OSX path strangeness
* [Spec] Attempt at fixing osx spec failure
* [Spec] More informative waits in core:save-as spec
* Fix specs

## 2.4.1
* Prepare 2.4.1 release
* Update CHANGELOG
* Merge branch 'faku99-master'
* Set engines to >=1.13.0
* Fix for issue #242 (Lucas Elisei)

## 2.4.0
* Prepare 2.4.0 release (Edward Fauchon-Jones)
* Merge pull request #196 from Galadirith/trim-trail-nl-only (Edward Fauchon-Jones)
* Fix linting error in lib/renderer (Edward Fauchon-Jones)
* Add new spec for triming trailing newline only (Edward Fauchon-Jones)
* Trim only trailing newline fixes #165 (Edward Fauchon-Jones)
* Change repo and version for markdown-it-math (Edward Fauchon-Jones)
* Bump markdown-it@5.1.0 (Edward Fauchon-Jones)
* Merge pull request #193 from Galadirith/jl-uncaught-string (Edward Fauchon-Jones)
* Fix failing spec due to leaking watched paths (Edward Fauchon-Jones)
* Move backwards-compatible deserializers.add call back into activate (Max Brunsfeld)
* Use new package.json fields to allow deferred loading (Max Brunsfeld)
* Add spec for not deserializing preview for deleted file (Kevin Sawicki)
* Require fs (Jessica Lord)
* Simplify checks per comments (Jessica Lord)
* Default to filePath (Jessica Lord)
* No previews unless id or path (Jessica Lord)
* Verify path is string before reading (Jessica Lord)

## 2.3.0
* Prepare 2.3.0 release (Edward Fauchon-Jones)
* Update package description (toned down) (Edward Fauchon-Jones)
* Update README for move to maintenance mode (Edward Fauchon-Jones)
* Merge pull request #192 from Galadirith/themed-preview (Edward Fauchon-Jones)
* Add list bottom margin (simurai)
* Adjust header size/weight (simurai)
* Use variables (simurai)
* Override link color (simurai)
* Prevent code color from getting overriden (simurai)
* Change padding (simurai)
* Simplify default styles (simurai)
* Split github and default styles into separate files (simurai)
* Show code scrollbars on hover (simurai)
* Remove scrollbar on code blocks (simurai)
* Merge pull request #171 from alerque/patch-1 (Edward Fauchon-Jones)
* Add text.md as a default grammar fixes #168 (Edward Fauchon-Jones)
* Merge pull request #191 from fix-style-init (Edward Fauchon-Jones)
* Merge pull request #184 from laokaplow/patch-1 (Edward Fauchon-Jones)
* Revert refactor of grammer assignment spec (Edward Fauchon-Jones)
* Correct bad refactor of grammer assignment spec (Edward Fauchon-Jones)
* Split js grammer assignment into own spec (Edward Fauchon-Jones)
* Remove redundant CI configurations (Edward Fauchon-Jones)
* Use ES6 .then instead of depreciated .done (Edward Fauchon-Jones)
* Update travis to use atom/ci configuration (Edward Fauchon-Jones)
* Update appveyor to use atom/ci configuration (Edward Fauchon-Jones)
* Update appveryor to use latest stable Atom release (Edward Fauchon-Jones)
* Update to fix coffeelint errors (Edward Fauchon-Jones)
* Update spec for failing toggle preview activation (Edward Fauchon-Jones)
* Update failing save and copy specs (Edward Fauchon-Jones)
* Move deserializer to package activation method (Edward Fauchon-Jones)
* Update `<atom-styles>` initialization (Edward Fauchon-Jones)
* Use new promise api (laokaplow)
* Fix MD syntax glitch for Atom's internal parser (Caleb Maclennan)
* Updated Changelog (Lukas Eipert)

## 2.2.2

src/config.ts
export const config = {
  breakOnSingleNewline: {
    type: 'boolean',
    default: false,
    order: 0,
  },
  liveUpdate: {
    type: 'boolean',
    default: true,
    order: 10,
  },
  openPreviewInSplitPane: {
    type: 'boolean',
    default: true,
    order: 20,
  },
  previewSplitPaneDir: {
    title: 'Direction to load the preview in split pane',
    type: 'string',
    default: 'right' as 'right' | 'down',
    enum: ['down', 'right'],
    order: 25,
  },
  grammars: {
    type: 'array',
    default: [
      'source.gfm',
      'source.litcoffee',
      'text.html.basic',
      'text.md',
      'text.plain',
      'text.plain.null-grammar',
    ],
    order: 30,
    items: {
      type: 'string',
    },
  },
  extensions: {
    type: 'array',
    title: 'Markdown file extensions',
    description: 'Which files are considered Markdown',
    default: ['markdown', 'md', 'mdown', 'mkd', 'mkdown', 'ron', 'txt'],
    order: 31,
    items: {
      type: 'string',
    },
  },
  enableLatexRenderingByDefault: {
    title: 'Enable Math Rendering By Default',
    type: 'boolean',
    default: false,
    order: 40,
  },
  latexRenderer: {
    title: 'Math Renderer',
    description:
      'SVG is noticeably faster, but might look worse on some systems',
    type: 'string',
    enum: ['HTML-CSS', 'SVG'],
    default: 'SVG' as 'HTML-CSS' | 'SVG',
    order: 41,
  },
  useLazyHeaders: {
    title: 'Use Lazy Headers',
    description: 'Require no space after headings #',
    type: 'boolean',
    default: true,
    order: 45,
  },
  useCheckBoxes: {
    title: 'Enable CheckBox lists',
    description: 'CheckBox lists, like on GitHub',
    type: 'boolean',
    default: true,
    order: 46,
  },
  useEmoji: {
    title: 'Use Emoji',
    description: 'Emoji rendering',
    type: 'boolean',
    default: true,
    order: 47,
  },
  inlineMathSeparators: {
    title: 'Inline math separators',
    description:
      'List of inline math separators in pairs -- first opening, then closing',
    type: 'array',
    default: ['$', '$', '\\(', '\\)'],
    order: 48,
    items: {
      type: 'string',
    },
  },
  blockMathSeparators: {
    title: 'Block math separators',
    description:
      'List of block math separators in pairs -- first opening, then closing',
    type: 'array',
    default: ['$$', '$$', '\\[', '\\]'],
    order: 48.1,
    items: {
      type: 'string',
    },
  },
  useGitHubStyle: {
    title: 'Use GitHub.com style',
    type: 'boolean',
    default: false,
    order: 50,
  },
  enablePandoc: {
    type: 'boolean',
    default: false,
    title: 'Enable Pandoc Parser',
    order: 100,
  },
  useNativePandocCodeStyles: {
    type: 'boolean',
    default: false,
    description: `\
Don't convert fenced code blocks to Atom editors when using
Pandoc parser`,
    order: 105,
  },
  pandocPath: {
    type: 'string',
    default: 'pandoc',
    title: 'Pandoc Options: Path',
    description: 'Please specify the correct path to your pandoc executable',
    dependencies: ['enablePandoc'],
    order: 110,
  },
  pandocFilters: {
    type: 'array',
    default: [] as string[],
    title: 'Pandoc Options: Filters',
    description:
      'Comma separated pandoc filters, in order of application. Will be passed via command-line arguments',
    dependencies: ['enablePandoc'],
    order: 115,
    items: {
      type: 'string',
    },
  },
  pandocArguments: {
    type: 'array',
    default: [] as string[],
    title: 'Pandoc Options: Commandline Arguments',
    description:
      'Comma separated pandoc arguments e.g. `--smart, --filter=/bin/exe`. Please use long argument names.',
    dependencies: ['enablePandoc'],
    order: 120,
    items: {
      type: 'string',
    },
  },
  pandocMarkdownFlavor: {
    type: 'string',
    default: 'markdown-raw_tex+tex_math_single_backslash',
    title: 'Pandoc Options: Markdown Flavor',
    description: 'Enter the pandoc markdown flavor you want',
    dependencies: ['enablePandoc'],
    order: 130,
  },
  pandocBibliography: {
    type: 'boolean',
    default: false,
    title: 'Pandoc Options: Citations',
    description: `\
Enable this for bibliography parsing.
Note: pandoc-citeproc is applied after other filters specified in
Filters, but before other commandline arguments\
`,
    dependencies: ['enablePandoc'],
    order: 140,
  },
  pandocRemoveReferences: {
    type: 'boolean',
    default: true,
    title: 'Pandoc Options: Remove References',
    description: 'Removes references at the end of the HTML preview',
    dependencies: ['pandocBibliography'],
    order: 150,
  },
  pandocBIBFile: {
    type: 'string',
    default: 'bibliography.bib',
    title: 'Pandoc Options: Bibliography (bibfile)',
    description: 'Name of bibfile to search for recursively',
    dependencies: ['pandocBibliography'],
    order: 160,
  },
  pandocBIBFileFallback: {
    type: 'string',
    default: '',
    title: 'Pandoc Options: Fallback Bibliography (bibfile)',
    description: 'Full path to fallback bibfile',
    dependencies: ['pandocBibliography'],
    order: 165,
  },
  pandocCSLFile: {
    type: 'string',
    default: 'custom.csl',
    title: 'Pandoc Options: Bibliography Style (cslfile)',
    description: 'Name of cslfile to search for recursively',
    dependencies: ['pandocBibliography'],
    order: 170,
  },
  pandocCSLFileFallback: {
    type: 'string',
    default: '',
    title: 'Pandoc Options: Fallback Bibliography Style (cslfile)',
    description: 'Full path to fallback cslfile',
    dependencies: ['pandocBibliography'],
    order: 175,
  },
}

// generated by typed-config.js
declare module 'atom' {
  interface ConfigValues {
    'markdown-preview-plus.breakOnSingleNewline': boolean
    'markdown-preview-plus.liveUpdate': boolean
    'markdown-preview-plus.openPreviewInSplitPane': boolean
    'markdown-preview-plus.previewSplitPaneDir': 'down' | 'right'
    'markdown-preview-plus.grammars': string[]
    'markdown-preview-plus.extensions': string[]
    'markdown-preview-plus.enableLatexRenderingByDefault': boolean
    'markdown-preview-plus.latexRenderer': 'HTML-CSS' | 'SVG'
    'markdown-preview-plus.useLazyHeaders': boolean
    'markdown-preview-plus.useCheckBoxes': boolean
    'markdown-preview-plus.useEmoji': boolean
    'markdown-preview-plus.inlineMathSeparators': string[]
    'markdown-preview-plus.blockMathSeparators': string[]
    'markdown-preview-plus.useGitHubStyle': boolean
    'markdown-preview-plus.enablePandoc': boolean
    'markdown-preview-plus.useNativePandocCodeStyles': boolean
    'markdown-preview-plus.pandocPath': string
    'markdown-preview-plus.pandocFilters': string[]
    'markdown-preview-plus.pandocArguments': string[]
    'markdown-preview-plus.pandocMarkdownFlavor': string
    'markdown-preview-plus.pandocBibliography': boolean
    'markdown-preview-plus.pandocRemoveReferences': boolean
    'markdown-preview-plus.pandocBIBFile': string
    'markdown-preview-plus.pandocBIBFileFallback': string
    'markdown-preview-plus.pandocCSLFile': string
    'markdown-preview-plus.pandocCSLFileFallback': string
    'markdown-preview-plus': {
      breakOnSingleNewline: boolean
      liveUpdate: boolean
      openPreviewInSplitPane: boolean
      previewSplitPaneDir: 'down' | 'right'
      grammars: string[]
      extensions: string[]
      enableLatexRenderingByDefault: boolean
      latexRenderer: 'HTML-CSS' | 'SVG'
      useLazyHeaders: boolean
      useCheckBoxes: boolean
      useEmoji: boolean
      inlineMathSeparators: string[]
      blockMathSeparators: string[]
      useGitHubStyle: boolean
      enablePandoc: boolean
      useNativePandocCodeStyles: boolean
      pandocPath: string
      pandocFilters: string[]
      pandocArguments: string[]
      pandocMarkdownFlavor: string
      pandocBibliography: boolean
      pandocRemoveReferences: boolean
      pandocBIBFile: string
      pandocBIBFileFallback: string
      pandocCSLFile: string
      pandocCSLFileFallback: string
    }
  }
}
